python/example_code/cloudwatch-logs/scenarios/large-query/cloudwatch_query.py

```diff
@@ -0,0 +1,193 @@
+        self.cloudwatch_logs = client
+        self.log_groups = log_groups
+        self.date_range = date_range
+        self.limit = limit
+        self.query_results = []
+        self.query_duration = None
+        self.datetime_format = "%Y-%m-%d %H:%M:%S.%f"
+        self.log_batch = None
+        self.date_utilities = DateUtilities()
+
+    def query_logs(self):
+        """
+        Executes a CloudWatch logs query for a specified date range and calculates the execution time of the query.
+
+        :return: A batch of logs retrieved from the CloudWatch logs query.
+        :rtype: list
+        """
+        start_time = datetime.now()
+
+        start_date, end_date = self.date_utilities.normalize_date_range_format(
+            self.date_range, from_format="unix_timestamp", to_format="datetime"
+        )
+
+        logging.info(
+            f"Original query:"
+            f"\n       START:    {start_date}"
+            f"\n       END:      {end_date}"
+        )
+        logs_batch = self.process_logs((start_date, end_date))
+        end_time = datetime.now()
+        self.query_duration = (end_time - start_time).total_seconds()
+        return logs_batch
+
+    def process_logs(self, date_range):
+        """
+        Processes logs within a given date range, fetching batches of logs recursively if necessary.
+
+        :param date_range: The date range to fetch logs for, specified as a tuple (start_timestamp, end_timestamp).
+        :type date_range: tuple
+        :return: None if the recursive fetching is continued or stops when the final batch of logs is processed.
+                 Although it doesn't explicitly return the query results, this method accumulates all fetched logs
+                 in the `self.query_results` attribute.
+        :rtype: None
+        """
+        # Fetch a batch of logs
+        batch_of_logs = self._perform_query(date_range, self.limit)
+        # Add the batch to the accumulated logs
+        self.query_results.extend(batch_of_logs)
+        logging.info(f"Recursive log count: {len(self.query_results)}")
+
+        # If the batch size is exactly 10,000, assume there might be more logs and fetch again
+        if len(batch_of_logs) == self.limit:
+            logging.info(f"Fetched {self.limit}, checking for more...")
+            most_recent_log = self._find_most_recent_log(batch_of_logs)
+            most_recent_log_timestamp = next(
+                item["value"]
+                for item in most_recent_log
+                if item["field"] == "@timestamp"
+            )
+            new_range = (most_recent_log_timestamp, date_range[1])
+            return self.process_logs(new_range)
+        else:
+            # If the batch size is less than 10,000, assume this is the last batch and return the accumulated logs
+            logging.info(f"Fetched final batch of {len(batch_of_logs)} logs.")
+
+    def _find_most_recent_log(self, logs):
```

Comment:
I did this too (my bad), but we probably don't need to "find most recent log" if our batch of logs is sorted ascending. Just grab the last log off the batch.